core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -537,7 +538,8 @@ class ReplicaManager(val config: KafkaConfig,
           val initialLogEndOffset = localReplica.logEndOffset
           val logReadInfo = localReplica.log match {
             case Some(log) =>
-              log.read(offset, fetchSize, maxOffsetOpt)
+              val adjustedFetchSize = if (TopicConstants.INTERNAL_TOPICS.contains(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
```

Comment:
If we're going to loosen this restriction for internal topics, why even have the restriction for other normal topics? This is basically saying that as long as some other broker said it was ok to write a larger messages than the limit we've placed on replication, we'll respect it. Seems like if we're going to do this, we might as well simplify things for normal topics that have been configured to accept larger messages so users don't need to override the fetch size setting.

In other words, are internal topics really special in this regard or have we just been making operators jump through annoying hoops when they have topics with larger max message sizes but really we just want to adjust to the max message size when needed?

Additionally, are we going to apply the same treatment to the consumers if we generalized? They could run into the same issue (and I've had complaints on kafka-rest about this behavior being implemented the same way for the REST proxy). The problem is that they don't have the relevant max message size info to figure out how to adjust their message size, although fetch request handlers on the broker could make the adjustment for them.